libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,637 @@
+      __derived_from_pack<typename iterator_traits<iterator_t<__maybe_const<_Const, _Views>>>::iterator_category...,
+                          bidirectional_iterator_tag>;
+  constexpr static bool derive_pack_forward_iterator =
+      __derived_from_pack<typename iterator_traits< iterator_t<__maybe_const<_Const, _Views>>>::iterator_category...,
+                          forward_iterator_tag>;
+  using iterator_category =
+      _If<!is_reference_v<__concat_reference_t<__maybe_const<_Const, _Views>...>>,
+          input_iterator_tag,
+          _If<derive_pack_random_iterator,
+              random_access_iterator_tag,
+              _If<derive_pack_bidirectional_iterator,
+                  bidirectional_iterator_tag,
+                  _If<derive_pack_forward_iterator, forward_iterator_tag, input_iterator_tag > > > >;
+  using iterator_concept =
+      _If<__concat_is_random_access<_Const, _Views...>,
+          random_access_iterator_tag,
+          _If<__concat_is_bidirectional<_Const, _Views...>,
+              bidirectional_iterator_tag,
+              _If< __all_forward<_Const, _Views...>, forward_iterator_tag, input_iterator_tag > > >;
+  using value_type      = __concat_value_t<__maybe_const<_Const, _Views>...>;
+  using difference_type = common_type_t<range_difference_t<__maybe_const<_Const, _Views>>...>;
+  using __base_iter     = variant<iterator_t<__maybe_const<_Const, _Views>>...>;
+
+  __base_iter it_;
+  __maybe_const<_Const, concat_view>* parent_ = nullptr;
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __satisfy() {
+    if constexpr (_Idx < (sizeof...(_Views) - 1)) {
+      if (std::get<_Idx>(it_) == ranges::end(std::get<_Idx>(parent_->__views_))) {
+        it_.template emplace<_Idx + 1>(ranges::begin(std::get<_Idx + 1>(parent_->__views_)));
+        __satisfy<_Idx + 1>();
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __prev() {
+    if constexpr (_Idx == 0) {
+      --std::get<0>(it_);
+    } else {
+      if (std::get<_Idx>(it_) == ranges::begin(std::get<_Idx>(parent_->__views_))) {
+        using __prev_view = __maybe_const<_Const, tuple_element_t<_Idx - 1, tuple<_Views...>>>;
+        if constexpr (common_range<__prev_view>) {
+          it_.template emplace<_Idx - 1>(ranges::end(std::get<_Idx - 1>(parent_->__views_)));
+        } else {
+          it_.template emplace<_Idx - 1>(ranges::__next(ranges::begin(std::get<_Idx - 1>(parent_->__views_)),
+                                                        ranges::size(std::get<_Idx - 1>(parent_->__views_))));
+        }
+        __prev<_Idx - 1>();
+      } else {
+        --std::get<_Idx>(it_);
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __advance_fwd(difference_type __offset, difference_type __steps) {
+    using underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, __base_iter>>;
```

Comment:
Missed __uglification, ditto below. (Luckily, it seems that we don't need to add `_LIBCPP_NODEBUG` to function-local typedefs.)
```suggestion
    using __underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, __base_iter>>;
```